dxvk_imgui.cpp - Lightmap/Hide Instance in UI

anchorlightforge opened the pull request:
*Note: An earlier commit was submitted from a different branch of the fork; this has now been remedied.*

Added two/three tags that were missing from the Texture UI but supported by Remix:
- Lightmap texture (optional)
- Hide Instance
- World Space UI Background Texture (commented out)

Due to the different behaviors of Lightmap Texture and Hide Instance I recommend exposing them to users to potentially improve compatibility in cases where hiding lightmaps as Ignore Textures causes additional rendering issues.

Deny this commit if hiding was intended.

pkristof reviewed (changes requested):
This looks good. Before this can be merged, following items should be addressed:
- expose the added texture categories in "Split Texture Category List" GUI view
- Remove or uncomment the `worldspaceuibackgroundtextures` item

If you are keeping `worldspaceuibackgroundtextures`, do following:
- add gui for worldSpaceUiBackgroundOffset
- add a description for `worldspaceuibackgroundtextures` in the RTX_OPTION. It's the last string in its definition. See a comment explaining it for when the offset is applied in rtx_instance_manager.cpp for description. The comment can be taken verbatim but strip the Portal reference and make the explanation more generically applicable.

anchorlightforge commented:
Sounds good! I read over your message earlier today.  I've made a few of these improvements already on a private branch, others should be relatively easy to implement.  I'll get right on this.

anchorlightforge commented:
I've incorporated your feedback, and opted to include World Space UI Background Textures.  I also noticed that Legacy Cutout Textures have a slider that until now was hidden in the original GUI, so I've opted to move this into a more accessible place as well.

In this PR I dropped the Step 1/Step 2 titles from the Categorize Textures and Parameter Tuning fields, only using them in the Split Texture GUI.  If desired I can revert that change.

The description I settled on for World Space UI Background Textures is as follows:

> Hack/workaround option for dynamic world space UI textures with a coplanar background.
> Apply to backgrounds if the foreground material is a dynamic world texture rendered in UI that is unpredictable and rapidly changing.
> This offsets the background texture backwards.

pkristof commented:
> The lack of an indent is bothering me a bit so I'll be putting together two alternatives with builds: one where the main "Categorize Textures" header disappears when the legacy GUI is enabled, and one where the indent is present. They should be available in a few hours when I'm back on my computer.

Let's preserve the `Categorize Textures` header as it encapsulates the texture tagging as a step even when all the lists are enumerated. Also preserve the 'Step ...' prefixes for the two main categories as it guides users to do the two steps as part of game setup. You can strip 'Step ...' prefix from the texture category lists on split view. The steps don't apply there anymore - the number of categories expanded dramatically since when the steps were introduced for the texture tagging as well as they are not present in the combined tagging view.

I do agree on adding an indent to `Categorize Textures` making it a bit better categorized visually.

anchorlightforge commented:
I am currently looking into implementing this indent.  However, adding an indent before the split texture fields is causing a unique problem-- any filtered list is somehow pushing the grid to the right at about the size of an indent, and pushing the entire list further to the right.  This behavior isn't present in the unfiltered list.  This also pushes a few of the textures past the scroll bar at the default size.  This was an issue before in the Sky Parameters field, but in order to cleanly implement this I'll have to stop ignoring the problem and figure out a way to fix it.

This only affects the selection grid, other UI elements inside these categories appear to be fine.  I'll keep investigating what I can do to make sure the split textures GUI displays correctly when indented like this.

![misaligned_grid](https://github.com/NVIDIAGameWorks/dxvk-remix/assets/55258193/7bfacb8e-42a9-4d2f-847b-cf9abc49c07d)

anchorlightforge commented:
An actions build of the current version is available here: https://github.com/anchorlightforge/dxvk-remix/actions/runs/5736793214

Since the last post I've addressed the following:
- Fixed incorrect bounds on the DragFloats added
- Added a slider and RtxOption that controls the scale of thumbnails with a boundary of 0.25x to 3x (compared to default of 120)
- Added indent to Split Texture category
- Added unindent and indent around all filtered list grids

pkristof commented:
Thanks for resolving those. Two more things:
- Can you squash all your commits into one for cleaner history?
- Add yourself to the "Github Contributors" list in ImGuiAbout::Credits::Credits(). It's sorted alphabetically by the last name.

pkristof commented on src/dxvk/rtx_render/rtx_options.h:
```diff
@@ -1236,6 +1242,7 @@ namespace dxvk {
     uint32_t getNumFramesToPutLightsToSleep() const { return numFramesToKeepLights() /2; }
     float getMeterToWorldUnitScale() const { return 100.f * getSceneScale(); } // RTX Remix world unit is in 1cm 
     float getSceneScale() const { return sceneScale(); }
+    float getTextureGridThumbnailScale() const { return textureGridThumbnailScale(); }
```
Remove this function and directly use `textureGridThumbnailScale()`

pkristof replied:
You can reference it as `RtxOptions::textureGridThumbnailScale()`

pkristof commented:
Sorry, I thought I reviewed your latest changes, let me review them before doing the squash.

pkristof commented on src/dxvk/imgui/dxvk_imgui.cpp:
```diff
@@ -1748,13 +1800,17 @@ namespace dxvk {
-
-      ImGui::DragFloat("Unique Object Search Distance", &RtxOptions::Get()->uniqueObjectDistanceObject(), 0.01f, 0.01f, FLT_MAX, "%.3f", sliderFlags);
       ImGui::Separator();
+      ImGui::DragFloat("Unique Object Search Distance", &RtxOptions::Get()->uniqueObjectDistanceObject(), 0.01f, 0.01f, FLT_MAX, "%.3f", sliderFlags);
```
Why was this moved? Unique Object Search Distance and Vertex Color Strength are unrelated and, thus, keeping them separated is intended.

pkristof commented on src/dxvk/imgui/dxvk_imgui.cpp:
```diff
@@ -1748,13 +1800,17 @@ namespace dxvk {
       ImGui::DragFloat("Vertex Color Strength", &RtxOptions::Get()->vertexColorStrengthObject(), 0.001f, 0.0f, 1.0f);
       ImGui::Separator();
-
+      if (ImGui::CollapsingHeader("Texture Parameters", collapsingHeaderClosedFlags)) {
```
Please remove duplicate widget entries listed in actual split texture category list that you added here. While it is somewhat useful having it in the actual texture category it does add to code maintenance cost and adds a risk of their parameterization getting divergent on any changes that forget to update both.

anchorlightforge commented:
I read over your corrections yesterday, and I'll be able to make the changes either tonight or tomorrow.  After that I'll squash them all into a new PR and close this one.  Thank you for your patience.

pkristof commented:
I don't manually squash but I imagine you should be able to squash them
within the same branch and just push that to the same branch. I will be
cherry picking the squashed commit to an internal branch and merging from
there so it is fine if the previous commits still show up in your branch.
The point of you squashing them is to preserve you being the author of
these improvements. Let's avoid recreating another PR if possible.

On Thu, Aug 3, 2023 at 9:50 AM Nico ***@***.***> wrote:

> I read over your corrections yesterday, and I'll be able to make the
> changes either tonight or tomorrow. After that I'll squash them all into a
> new PR and close this one. Thank you for your patience.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIAGameWorks/dxvk-remix/pull/30#issuecomment-1664231710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGF4CIPDAQYGZJZMF5IGZETXTPCFFANCNFSM6AAAAAAZJ7FUQY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

anchorlightforge commented:
I was hoping to get back to this over the weekend since I didn't have as much free time Friday as I expected, but unfortunately it doesn't seem like that'll happen.  I'll try to get around to completing the changes by Thursday-- it should be pretty much identical to your feedback, which are mostly down to either get-it-work mistakes or misunderstandings that I've taken note of for documentation's sake.  There is one more feature I want to implement where you can have a single texture panel and a combo box that determines what will happen when you click, since just as in the original legacy implementation the split texture boxes can easily become cluttered and confusing to sift through.  I'll update this thread if there are any changes before then.

anchorlightforge commented:
Thank you for your extended patience.  I've committed your review changes and will squash-merge them as soon as the build test completes.

pkristof commented:
Your last commit is not squashed with all and only your changes. Given the fact that your commits contain changes from dxvk-remix, I think the best course of action here is:
1. Rebase/merge your PR's branch onto latest `dxvk-remix`. 
2. Create a new branch off the latest `dxvk-remix`.
3. **Squash and merge** PR's branch into the new branch with an appropriate commit message concisely describing what the change does (i.e. "Extended texture tagging UI").
4. [Change the PR's base branch to the new one. ](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request)
5. Make sure the PR shows just the commit with only your changes

That should result in a single commit with just and all of your changes.

anchorlightforge commented:
The rebasing is complete and down to one commit on https://github.com/anchorlightforge/dxvk-remix/tree/PR30-upstream-1 now.

I'm having difficulty changing the base branch as the only available option is NVIDIAGameWorks:main.  What would be the best way to remedy this without creating a new branch?

https://github.com/NVIDIAGameWorks/dxvk-remix/commit/9275952ee8cd3ea7a30c717e74342cebfb798d7c

pkristof commented:
Apologies, the link I shared describes changing of the base branch for a PR, not the topic branch that you want to merge into the base branch. If the PR settings don't let you change the topic branch, then the other option is to force-push your new branch into your existing branch that you created the PR from. If that doesn't work out and the PR doesn't show the history as that single commit, then please close this PR and cross link the PRs by adding the links into their description.

anchorlightforge commented:
Phew, seems like it's all set!  Thanks again for walking me through the merge process.  I'm currently typing up a community post to help others hopefully avoid the same pitfalls I fell into working on this PR, and at some point this week I'll give a pass to the other PRs I have in progress to make sure they're cleaned up with a single commit.

pkristof commented:
Good to hear it worked out. I think writing instructions & gotchas on PRs is a great idea. That's something that could be added as `Contibuting.md` file into the repo and linked to from the main Readme.md.

anchorlightforge commented:
Sounds good. I'm drafting the post in the #open-source-projects section of the community Discord, then I'll make a separate PR for the Contributing file later down the line.

pkristof commented on src/dxvk/imgui/dxvk_imgui.cpp:
```diff
@@ -1636,108 +1638,158 @@ namespace dxvk {
+          ImGui::Indent();
         }
 
-        if (IMGUI_ADD_TOOLTIP(ImGui::CollapsingHeader("Step 7.2: Single Offset Decal Textures", collapsingHeaderClosedFlags), RtxOptions::Get()->singleOffsetDecalTexturesDescription())) {
```
Add back this category `singleoffsetdecaltextures` (I suspect this is fallout from a merge conflict).

pkristof commented on src/dxvk/imgui/dxvk_imgui.cpp:
```diff
@@ -1748,13 +1800,16 @@ namespace dxvk {
       ImGui::Checkbox("Scene Left-Handed", &RtxOptions::Get()->isLHSObject());
       fusedWorldViewModeCombo.getKey(&RtxOptions::Get()->fusedWorldViewModeRef());
       ImGui::Separator();
-
```
Revert the empty line changes. Generally should avoid making any unnecessary/unwanted changes as it complicates the PR.

pkristof commented:
I have left you two comments. It's fine to address them in a 2nd commit - it's acceptable as far as commit history goes but if you really wanted you could squash it again.

pkristof commented:
This has been merged. Thank you for the contribution. This PR can be closed.

anchorlightforge commented:
Thank you again for working with me through this! This ended up being a learning experience for me. I've drafted a contributing.md locally and will put up a PR draft for it later this week.

pkristof commented:
Sounds good, please @at me on that PR and I will review it. I know that you have a few PRs open that we have not had a chance to respond to. Sorry about that - there is just a ton of things in flight.

MarkEHenderson commented:
> Thank you again for working with me through this! This ended up being a learning experience for me. I've drafted a contributing.md locally and will put up a PR draft for it later this week.

@anchorlightforge Did you ever get around to writing up that contributing.md file?

anchorlightforge commented:
@MarkEHenderson There was some poking around with it in the Discord, most likely buried at this point.  It was written up locally as well but needed some polishing-- it was written like a FAQ but probably needs some rearrangements.  It's on a stale branch somewhere but I won't have a chance to dig it back up until the weekend.
